format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java

```diff
@@ -141,11 +142,57 @@ public static void validateSchemaFieldNames(String[] fieldNames) {
         throw new IllegalArgumentException(
           String.format(
             "Invalid column name detected: \"%s\"! " +
-            "Column names can only start with capital letters, small letters, and " + "\"_\". " +
-            "Subsequently they may contain only capital letters, small letters, numbers  and \"_\".", fieldName
+              "Column names can only start with capital letters, small letters, and " + "\"_\". " +
+              "Subsequently they may contain only capital letters, small letters, numbers  and \"_\".", fieldName
           )
         );
       }
     }
   }
+
+  /**
+   * Cleans an array of field names to make sure they comply with avro field naming standard.
+   * It also makes sure each name is unique in the list.
+   * Field names can start with [A-Za-z_] and subsequently contain only [A-Za-z0-9_].
+   *
+   * Steps:
+   * 1) Trim surrounding spaces
+   * 2) If it starts with a number, prepend "_"
+   * 3) Replace invalid characters with "_"
+   * 4) Check if the name has been found before if so add _# where # is the number of times seen before
+   * @param fieldNames an array of field names to be cleaned
+   */
+  private static String[] cleanSchemaFieldNames(String[] fieldNames) {
+    final String replacementChar = "_";
+    final List<String> cleanFieldNames = new ArrayList<>();
+    final Map<String, Integer> seenFieldNames = new HashMap<>();
+    final Pattern notValidPattern = Pattern.compile("[^A-Za-z0-9_]");
+    for (String fieldName : fieldNames) {
+      StringBuilder cleanFieldNameBuilder = new StringBuilder();
+
+      // Remove any spaces at the end of the strings
+      fieldName = fieldName.trim();
+
+      // If it starts with a number, or it's an empty string prepend the replacement char
+      if ((fieldName.length() == 0) || (Character.isDigit(fieldName.charAt(0)))) {
+        cleanFieldNameBuilder.append(replacementChar);
+      }
+
+      // Replace all invalid characters with the replacement char
+      cleanFieldNameBuilder.append(notValidPattern.matcher(fieldName).replaceAll(replacementChar));
+
+      // Check if the field exist if so append and index at the end
+      String cleanFieldName = cleanFieldNameBuilder.toString();
+
+      while (seenFieldNames.containsKey(cleanFieldName)) {
+        cleanFieldNameBuilder.append(replacementChar).append(seenFieldNames.get(cleanFieldName));
+        seenFieldNames.put(cleanFieldName, seenFieldNames.get(cleanFieldName) + 1);
+        cleanFieldName = cleanFieldNameBuilder.toString();
+      }
+      seenFieldNames.put(cleanFieldName, 1);
```

Comment:
So, it gives us "a", "a_1", "a_2", correct?
Should it be "a", "a_2", "a_3"? "a" and "a_1" can be easily confused